hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java

```diff
@@ -524,41 +531,82 @@ private HoodieData<String> repartitioningIfNeeded(
+    Map<StoragePath, HoodieAvroFileReader> baseFileReaders = Collections.emptyMap();
+    ReusableFileGroupRecordBufferInitializer<IndexedRecord> recordBufferInitializer = null;
+    if (reuse && isFullScanAllowedForPartition(fileSlice.getPartitionPath())) {
+      Pair<HoodieAvroFileReader, ReusableFileGroupRecordBufferInitializer<IndexedRecord>> readers =
+          reusableFileReaders.computeIfAbsent(fileSlice.getFileGroupId(), fgId -> {
+            try {
+              HoodieAvroFileReader baseFileReader = null;
+              if (fileSlice.getBaseFile().isPresent()) {
+                baseFileReader = (HoodieAvroFileReader) HoodieIOFactory.getIOFactory(storage).getReaderFactory(HoodieRecord.HoodieRecordType.AVRO)
+                    .getFileReader(metadataConfig, fileSlice.getBaseFile().get().getStoragePath(), metadataMetaClient.getTableConfig().getBaseFileFormat(), Option.empty());
+              }
+              return Pair.of(baseFileReader, buildReusableRecordBufferInitializer(fileSlice, latestMetadataInstantTime, instantRange));
+            } catch (IOException ex) {
+              throw new HoodieIOException("Error opening readers for metadata table partition " + fileSlice.getPartitionPath(), ex);
+            }
+          });
+      if (fileSlice.getBaseFile().isPresent()) {
+        baseFileReaders = Collections.singletonMap(fileSlice.getBaseFile().get().getStoragePath(), readers.getLeft());
+      }
+
+      ValidationUtils.checkArgument(isFullKey, "For Metadata Table Reuse, key filter should be based on full keys");
+      recordBufferInitializer = readers.getRight();
+    }
+
     HoodieReaderContext<IndexedRecord> readerContext = new HoodieAvroReaderContext(
         storageConf,
         metadataMetaClient.getTableConfig(),
-        Option.of(instantRange),
+        instantRange,
         Option.of(
             isFullKey
                 ? transformKeysToPredicate(sortedKeys)
-                : transformKeyPrefixesToPredicate(sortedKeys)));
-    return HoodieFileGroupReader.<IndexedRecord>newBuilder()
+                : transformKeyPrefixesToPredicate(sortedKeys)),
+        baseFileReaders);
+
+    HoodieFileGroupReader<IndexedRecord> fileGroupReader = HoodieFileGroupReader.<IndexedRecord>newBuilder()
         .withReaderContext(readerContext)
         .withHoodieTableMetaClient(metadataMetaClient)
         .withLatestCommitTime(latestMetadataInstantTime)
         .withFileSlice(fileSlice)
-        .withDataSchema(schema)
-        .withRequestedSchema(schema)
+        .withDataSchema(SCHEMA)
+        .withRequestedSchema(SCHEMA)
         .withProps(buildFileGroupReaderProperties(metadataConfig))
-        .withStart(0)
-        .withLength(Long.MAX_VALUE)
-        .withShouldUseRecordPosition(false)
+        .withRecordBufferInitializer(recordBufferInitializer)
         .build();
+
+    return fileGroupReader.getClosableIterator();
```

Comment:
The iterator is always closing now?

Reply:
I can also override the close call and add an inline comment so it is clear what the expectation is